nemo/collections/asr/parts/submodules/multi_head_attention.py

```diff
@@ -240,18 +279,43 @@ def forward(self, query, key, value, mask, pos_emb, cache=None):
             # first compute matrix a and matrix c
             # as described in https://arxiv.org/abs/1901.02860 Section 3.3
             # (batch, head, time1, time2)
-            matrix_ac = torch.matmul(q_with_bias_u, k.transpose(-2, -1))
 
             # compute matrix b and matrix d
             # (batch, head, time1, time2)
             matrix_bd = torch.matmul(q_with_bias_v, p.transpose(-2, -1))
             matrix_bd = self.rel_shift(matrix_bd)
-            # drops extra elements in the matrix_bd to match the matrix_ac's size
-            matrix_bd = matrix_bd[:, :, :, : matrix_ac.size(-1)]
 
-            scores = (matrix_ac + matrix_bd) / self.s_d_k  # (batch, head, time1, time2)
-
-            out = self.forward_attention(v, scores, mask)
+            if self.use_pytorch_sdpa:
+                scale_factor = 1 / math.sqrt(q_with_bias_u.size(-1))
+                matrix_bd = matrix_bd[:, :, :, : k.size(-2)] * scale_factor
+
+                if mask is not None:
+                    mask = mask.unsqueeze(1)
+                    matrix_bd.masked_fill_(mask, -10000.0)
+                    # add extra col for mask (matrix_bd) to handle problem with nan after solfmax
+                    rows_all_false = torch.all(mask, dim=-1)
+                    new_column = torch.where(rows_all_false, torch.tensor(10000.0), torch.tensor(-10000.0))
+                    new_column = new_column.repeat(1, self.h, 1)
+                    matrix_bd = torch.cat([matrix_bd, new_column.unsqueeze(-1)], dim=-1).to(matrix_bd.device)
+
+                dropout_rate = self.dropout_rate if self.training else 0
+                # add extra col for key and value to handle problem with nan after solfmax
+                extra_column = torch.zeros(k.shape[:-2] + (1, k.shape[-1])).to(k.device)
+                k = torch.cat([k, extra_column], dim=-2)
+
+                extra_column = torch.zeros(v.shape[:-2] + (1, v.shape[-1])).to(v.device)
+                v = torch.cat([v, extra_column], dim=-2)
+                out = torch.nn.functional.scaled_dot_product_attention(
```

Comment:
I just noticed: isn;t this different from `matrix_ac + matrix_bd` for computation of attention scores? How are both equal?

Reply:
Here is the SDPA code from the [pytorch documentation](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html).
I passed `matrix_bd` as `attn_mask`, which then becomes `attn_bias`:
```python
attn_bias = torch.zeros(L, S, dtype=query.dtype)
attn_bias += attn_mask
```

`attn_weight` is similar to `matrix_ac`:
```python
attn_weight = query @ key.transpose(-2, -1) * scale_factor
```

This line `attn_weight += attn_bias` is equivalent to `matrix_ac + matrix_bd`:
```python
# Efficient implementation equivalent to the following:
def scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=0.0, is_causal=False, scale=None) -> torch.Tensor:
    L, S = query.size(-2), key.size(-2)
    scale_factor = 1 / math.sqrt(query.size(-1)) if scale is none else scale
    attn_bias = torch.zeros(L, S, dtype=query.dtype)
    if is_causal:
        assert attn_mask is none
        temp_mask = torch.ones(L, S, dtype=torch.bool).tril(diagonal=0)
        attn_bias.masked_fill_(temp_mask.logical_not(), float("-inf"))
        attn_bias.to(query.dtype)

    if attn_mask is not none:
        if attn_mask.dtype == torch.bool:
            attn_bias.masked_fill_(attn_mask.logical_not(), float("-inf"))
        else:
            attn_bias += attn_mask
    attn_weight = query @ key.transpose(-2, -1) * scale_factor
    attn_weight += attn_bias
    attn_weight = torch.softmax(attn_weight, dim=-1)
    attn_weight = torch.dropout(attn_weight, dropout_p, train=True)
    return attn_weight @ value
```